### PR TITLE
Implementing User Storage Provider

### PR DESCRIPTION
**Pull request by prakharsaxena3007**

### CODE REVIEW

File: src/main/java/org/example/keycloak/SpringUserStorageProvider.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    public void close() { }
+
+    @Override
+    public Stream<UserModel> searchForUserStream(RealmModel realm, String search) {
+        return hashMapUserStore.findUsers(search).stream()
+                .map(user -> new UserAdapter(session, realm, model, user));
+    }
+
+    @Override
+    public Stream<UserModel> searchForUserStream(RealmModel realmModel, String s, Integer integer, Integer integer1) {
```
Comment (ankit-fiftyfive) [inline src/main/java/org/example/keycloak/SpringUserStorageProvider.java]: Replace String s with proper name

File: src/main/java/org/example/keycloak/UserApiService.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
+
+public class UserApiService {
+    public Logger logger = LoggerFactory.getLogger(UserApiService.class);
+    ObjectMapper objectMapper = new ObjectMapper();
+
+    public List<User> getAllUsers(String url) throws NullPointerException {
```
Comment (ankit-fiftyfive) [inline src/main/java/org/example/keycloak/UserApiService.java]: Please remove null pointer exception

File: src/main/java/org/example/keycloak/external/HashMapUserStore.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+    public HashMapUserStore(UserApiService userApiService){
+            this.userApiService = userApiService;
+    }
+
+    public List<User> getAllUsers() {
+        return users;
+    }
+
+    public User findUserById(String id) {
```
Comment (ankit-fiftyfive) [inline src/main/java/org/example/keycloak/external/HashMapUserStore.java:21]: Remove these print statements

File: pom.xml [hunk 4]
```diff
@@ -13,6 +13,69 @@
...
+        <dependency>
+            <groupId>org.keycloak</groupId>
+            <artifactId>keycloak-services</artifactId>
+            <version>21.1.1</version>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>com.mopano</groupId>
+            <artifactId>hibernate-json-org-contributor</artifactId>
+            <version>1.0</version>
```
Comment (raman7073) [inline pom.xml]: why are we using this dependency?

Comment (prakharsaxena3007) [reply pom.xml]: I forget to remove this dependency, I was using JSONObject and JSONArray  before using JSONNode which gave me error for NO class found. this dependency was for those class. I will remove it now as it is of no use now.

File: src/main/java/org/example/keycloak/UserApiService.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+            URL userApiUrl = new URL(url);
+
+            HttpURLConnection connection = (HttpURLConnection) userApiUrl.openConnection();
+            connection.setRequestMethod("GET");
+
+            int responseCode = connection.getResponseCode();
+
+            if (responseCode == HttpURLConnection.HTTP_OK) {
+                BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
+                String line;
```
Comment (ankit-fiftyfive) [inline src/main/java/org/example/keycloak/UserApiService.java]: Please provide more specific variable name

File: src/main/java/org/example/keycloak/UserApiService.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+                StringBuilder response = new StringBuilder();
+
+                while ((line = reader.readLine()) != null) {
```
Comment (ankit-fiftyfive) [inline src/main/java/org/example/keycloak/UserApiService.java]: Please use Objects.nonnull() method to check null values

File: src/main/java/org/example/keycloak/UserApiService.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+                        String username = jsonNode.has("username") ?jsonNode.get("username").asText():null;
+                        String firstName = jsonNode.has("first_name") ? jsonNode.get("first_name").asText() :null ;
+                        String lastName = jsonNode.has("last_name") ? jsonNode.get("last_name").asText() : null ;
+
+                        User user = new User(id,username,firstName,lastName);
+                        userList.add(user);
+                    }
+                }
+            } else {
+                logger.error("HTTP request failed with response code: " +responseCode);
```
Comment (ankit-fiftyfive) [inline src/main/java/org/example/keycloak/UserApiService.java]: Please use String formatter

File: src/main/java/org/example/keycloak/UserApiService.java [hunk 8]
```diff
@@ -30,10 +30,10 @@ public List<User> getAllUsers(String url){
 
             if (responseCode == HttpURLConnection.HTTP_OK) {
                 BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
-                String allUsersFromApi = reader.readLine();
+                String allUsersFromApi ;
                 StringBuilder response = new StringBuilder();
 
-                while (Objects.nonNull(allUsersFromApi)) {
```
Comment (prakharsaxena3007) [inline src/main/java/org/example/keycloak/UserApiService.java]: here this variable was not updating in the loop. That is why I have to instantiate it in while loop only. Before it was giving me error regarding heap memory Out of storage error.

**Review by raman7073 — APPROVED** [review]

**Pull request merged**